src/boot/efi/addon-util.c

```diff
@@ -0,0 +1,36 @@
+#include "addon-util.h"
+#include "proto/device-path.h"
+#include "util.h"
+
+static const struct {
+        VENDOR_DEVICE_PATH vendor;
+        EFI_DEVICE_PATH end;
+} _packed_ efi_addon_device_path = {
+        .vendor = {
+                .Header = {
+                        .Type = MEDIA_DEVICE_PATH,
+                        .SubType = MEDIA_VENDOR_DP,
+                        .Length = sizeof(efi_addon_device_path.vendor),
+                },
+                .Guid = SYSTEMD_ADDON_MEDIA_GUID
+        },
+        .end = {
+                .Type = END_DEVICE_PATH_TYPE,
+                .SubType = END_ENTIRE_DEVICE_PATH_SUBTYPE,
+                .Length = sizeof(efi_addon_device_path.end),
+        }
+};
+
+EFI_STATUS addons_install(EFI_HANDLE device, const char16_t **addons, EFI_HANDLE *ret_addons_handle)
+{
+        EFI_STATUS err;
+        EFI_DEVICE_PATH *addons_dp = (EFI_DEVICE_PATH *) &efi_addon_device_path;
+
+        err = make_multiple_file_device_path(device, addons, &addons_dp);
+        if (err != EFI_SUCCESS)
+                return err;
+
+        return BS->InstallMultipleProtocolInterfaces(ret_addons_handle,
+                                            MAKE_GUID_PTR(EFI_DEVICE_PATH_PROTOCOL),
+                                            &addons_dp);
+}
```

Comment:
You're thinking too much of how the initrd loadfile stuff works.

There is no need for a (global) efi_addon_device_path struct or a new handle for it.

Just create a heap allocated mulit device path and install it onto the handle of the *loaded image*:
```
BS->LoadImage(…, loaded_image_handle);
…
EFI_DEVICE_PATH *dp = make_multiple_file_device_path(…);
…
BS->InstallMultipleProtocolInterfaces(loaded_image_handle, MAKE_GUID_PTR(SYSTEMD_ADDON_DEVICE_PATH), &dp, NULL);
```

This way, the stub can just look at the handle it receives in efi_main for the device path and there is no guesswork to be done. Also, it has to be under our own new GUID as `EFI_DEVICE_PATH_PROTOCOL_GUID` as fixed meanings defined by the UEFI spec. We cannot step on that.

Reply:
Oh, I see, I was not sure, I could directly install it with my custom GUID, I can just keep the function right and pass the `loaded_image_handle` or should I just ensure I get an EFI loaded image handle to ensure no one confuses the handle?
Thank you for the pointer :)